lib/src/authentication/token.dart

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import '../source/hosted.dart';
+
+/// Token is a structure for storing authentication credentials for third-party
+/// pub registries. A token holds registry [url], credential [kind] and [token]
+/// itself.
+///
+/// Token could be serialized into and from JSON format structured like
+/// this:
+///
+/// ```json
+/// {
+///   "url": "https://example.com/",
+///   "credential": {
+///     "kind": "Bearer",
+///     "token": "gjrjo7Tm2F0u64cTsECDq4jBNZYhco"
+///   }
+/// }
+/// ```
+class Token {
+  Token({required this.url, required this.kind, required this.token});
+
+  /// Create [Token] instance with the `'Bearer'` kind.
+  Token.bearer(this.url, this.token) : kind = 'Bearer';
+
+  /// Deserialize [json] into [Token] type.
```

Comment:
Suggest adding:
```
///
/// Throws [FormatException] if [json] is not a valid [Token].
```

This is just a nit thing, since it's an internal method.
But it's nice to try to document which exceptions can be thrown, only `Exception`s, not
`Error`s as callers are not expected to handle them.

In this case, it can be important that the caller handles `FormatException`s somehow.